### PR TITLE
3P Media: Add search capabilities

### LINKED ISSUE #2391: Wire up and style Search bar widget
Widget should have a "X" that shows up on focus (and unfocuses the search bar on click).

Widget should be above the Unsplash tab.

The query string doesn't have to be stored in the state yet. (done in separate ticket)

Let's use the existing search bar component.

### LINKED ISSUE #2430: Add search term state to media/media3p state
## Task Description

Add 'searchTerm' to track the search term for media3p that's applied to the list.
When set, the media fetch side effect will automatically execute the search query to the server (see https://github.com/google/web-stories-wp/issues/2824).

Note that this state represents the search term that's applied, not necessarily what is in the UI search field.

* Add 'searchTerm' to media/media3p/reducer.js and useProviderContextValueProvider.js.
* Add SET_SEARCH_TERM to media/media3p/types.js and a dispatcher to media/media3p/actions.js.
* Add a reducer to set the searchTerm state.
* In the reducer, if the search term has changed, reset pageToken / nextPageToken to undefined so that results are loaded from the 1st page.

### PR DESCRIPTION
**Pull request by diegovar**
## Summary

Wires up search into the media3p tab

Screen Shot 2020-07-13 at 3 39 10 pm

## Relevant Technical Choices

None.

## To-do

Add karma test.
Possibly remove debouncing when the search term is empty (to not see a delay when clicking the X).

## User-facing changes

Search now works.

## Testing Instructions

Try out searching for images for given terms.

---

Fixes #2391

### CODE REVIEW

File: assets/src/edit-story/app/media/common/reducer.js [hunk 1]
```diff
@@ -56,10 +56,11 @@ function reducer(state = INITIAL_STATE, { type, payload }) {
     }
 
     case types.FETCH_MEDIA_SUCCESS: {
-      const { media, nextPageToken, totalPages } = payload;
+      const { media, pageToken, nextPageToken, totalPages } = payload;
       return {
         ...state,
-        media: [...state.media, ...media],
+        // If a pageToken is present, append the results.
+        media: pageToken ? [...state.media, ...media] : media,
```
Comment (cvolzke4) [inline assets/src/edit-story/app/media/common/reducer.js:63]: +1

File: assets/src/edit-story/app/media/media3p/actions.js [hunk 2]
```diff
@@ -26,3 +26,10 @@ export const setSelectedProvider = (dispatch) => ({ provider }) => {
     payload: { provider },
   });
 };
+
+export const search = (dispatch) => ({ searchTerm }) => {
```
Comment (cvolzke4) [inline assets/src/edit-story/app/media/media3p/actions.js]: Other action dispatchers are named as 'set*' etc when setting state, for example setMediaType, setProcessing and setSearchTerm in local.js.

Can this be named similarly, so setSearchTerm / types.SET_SEARCH_TERM?

File: assets/src/edit-story/app/media/media3p/reducer.js [hunk 3]
```diff
@@ -74,6 +75,12 @@ function reducer(state = INITIAL_STATE, { type, payload }) {
         selectedProvider: payload.provider,
       };
     }
+    case types.SEARCH: {
+      return {
+        ...state,
+        searchTerm: payload.searchTerm,
```
Comment (cvolzke4) [inline assets/src/edit-story/app/media/media3p/reducer.js:81]: This can also set 'pageToken: undefined' if the searchTerm has actually changed, so that fetching starts from the 1st page.

File: assets/src/edit-story/app/media/media3p/useFetchMediaEffect.js [hunk 4]
```diff
@@ -42,9 +43,10 @@ export default function useFetchMediaEffect({
       try {
         const { media, nextPageToken } = await listMedia({
           provider,
+          searchTerm,
           pageToken,
         });
-        fetchMediaSuccess({ provider, media, nextPageToken });
+        fetchMediaSuccess({ provider, media, pageToken, nextPageToken });
```
Comment (cvolzke4) [inline assets/src/edit-story/app/media/media3p/useFetchMediaEffect.js:49]: pageToken doesn't need to be passed to fetchMediaSuccess, only nextPageToken. The 'setNextPage' action is responsible for assigning pageToken = nextPageToken.

File: assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js [hunk 5]
```diff
@@ -127,13 +141,22 @@ function Media3pPane(props) {
       currentPaddingLeft - scrollbarWidth + 'px';
   }, [scrollbarWidth, refContainer]);
 
-  const { media } = useMedia3pForProvider('unsplash', ({ state }) => ({
-    media: state.media,
-  }));
+  /**
+   * Effectively performs a search, triggered at most every 500ms.
+   */
+  const [changeSearchTermDebounced] = useDebouncedCallback(() => {
```
Comment (cvolzke4) [inline assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js:105]: I thought that the search was only going to apply when the user presses enter.

File: assets/src/edit-story/app/media/common/reducer.js (hunk 1 above)
Comment (winterDroid) [reply assets/src/edit-story/app/media/common/reducer.js:63]: Should we cover this with a test?

File: assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js [hunk 6]
```diff
@@ -127,13 +141,22 @@ function Media3pPane(props) {
...
-    // TODO(#2391): Perform search.
-  }, []);
+  /**
+   * Handle search input changes. Triggers with every keystroke.
+   *
+   * @param {string} value the new search term.
+   */
+  const onSearch = (value) => {
+    setSearchTermValue(value);
+    changeSearchTermDebounced();
```
Comment (winterDroid) [inline assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js:130]: Wasn't this part of another PR? 🤔

File: assets/src/edit-story/app/media/media3p/useFetchMediaEffect.js (hunk 4 above)
Comment (diegovar) [reply assets/src/edit-story/app/media/media3p/useFetchMediaEffect.js:49]: Needs to be passed so we decide if we have to append or replace, right?

File: assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js (hunk 5 above)
Comment (diegovar) [reply assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js:105]: I've asked Sam for clarification, as the current upload tab searches automatically without enter.

File: assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js (hunk 6 above)
Comment (diegovar) [reply assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js:130]: It was for the upload tab. This PR doesn't yet unify those 2 things (it's repeated now).

File: assets/src/edit-story/app/media/common/reducer.js (hunk 1 above)
Comment (diegovar) [reply assets/src/edit-story/app/media/common/reducer.js:63]: Done.

File: .github/workflows/continuous-integration-karma-edit-story.yml [hunk 7]
```diff
@@ -76,7 +76,7 @@ jobs:
         uses: actions/checkout@v2
         with:
           ref: gh-pages
-  
+
```
Comment (cvolzke4) [inline .github/workflows/continuous-integration-karma-edit-story.yml:79]: [nit]: Is this change intended?

File: assets/src/edit-story/app/media/media3p/test/reducer.js [hunk 8]
```diff
@@ -71,4 +71,18 @@ describe('reducer', () => {
       })
     );
   });
+
+  it('should assign searchTerm on search', () => {
```
Comment (cvolzke4) [inline assets/src/edit-story/app/media/media3p/test/reducer.js]: "... on setSearchTerm"

File: assets/src/edit-story/app/media/media3p/useFetchMediaEffect.js (hunk 4 above)
Comment (cvolzke4) [reply assets/src/edit-story/app/media/media3p/useFetchMediaEffect.js:49]: That makes sense.

File: assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js (hunk 5 above)
Comment (cvolzke4) [reply assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js:105]: It's also an eng decision, if we're concerned about the qps sent to Unsplash / Coverr. This is a bigger concern for Coverr than Unsplash though, so we can decide on this later.

File: assets/src/edit-story/karma-tests.cjs [hunk 9]
```diff
@@ -15,5 +15,10 @@
  */
 
 // Collect all Karma tests together for the test webpack config.
-const testsContext = require.context('.', true, /\.karma\.js$/);
+// const testsContext = require.context('.', true, /\.karma\.js$/);
+const testsContext = require.context(
+  './components/library/panes/media/media3p',
```
Comment (cvolzke4) [inline assets/src/edit-story/karma-tests.cjs]: revert

**Review by cvolzke4 — APPROVED** [review]

**Review by winterDroid — APPROVED** [review]

File: .github/workflows/continuous-integration-karma-edit-story.yml (hunk 7 above)
Comment (diegovar) [reply .github/workflows/continuous-integration-karma-edit-story.yml:79]: Not necessarily but seems to be introduced by the linter, so I'd rather leave it (someone will have to merge it at some point.)

File: assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js (hunk 5 above)
Comment (diegovar) [reply assets/src/edit-story/components/library/panes/media/media3p/media3pPane.js:105]: Will remove autocompolete in a subsequent PR
